cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/generator/InputInterpreterImpl.java

```diff
@@ -68,23 +69,53 @@ public Object read(String type, Path path, Charset inputCharset, Object... addit
     public Object read(Path path, Charset inputCharset, Object... additionalArguments) throws InputReaderException {
         Set<String> keySet = PluginRegistry.getTriggerInterpreterKeySet();
         // We first try to find an input reader that is most likely readable
+        HashMap<String, Boolean> readableCache = new HashMap<String, Boolean>();
+
+        // Create cache for readable states
         for (String s : keySet) {
+            readableCache.put(s, isMostLikelyReadable(s, path));
+        }
+
+        // Check external input readers first
+        for (String s : readableCache.keySet()) {
+            try {
+                if (readableCache.get(s) != null && readableCache.get(s)) {
+                    LOG.info("Try reading input {} with EXTERNAL inputreader '{}'...", path, s);
+                    return getInputReader(s).read(path, inputCharset, additionalArguments);
+                }
+            } catch (InputReaderException e) {
+                LOG.debug(
+                    "Was not able to read input {} with EXTERNAL inputreader '{}' although it was reported to be most likely readable. Trying next input reader...",
+                    path, s, e);
+            } catch (Throwable e) {
+                LOG.debug(
+                    "While reading the input {} with the EXTERNAL inputreader {}, an Exception occured. Trying next input reader...",
+                    path, s, e);
+            }
+        }
```

Comment:
this and the below is the exact same code. I would like to externalize it to a common method, which takes two parameters: the readableCache and the value to compare to for activation (e.g. true or null).

It would be also better, if you don't query all isMostLikelyReadable in advance to you have them cached. You should build up the cache as soon as you need to query a new plugin. This will increase performance tremendously as you might in a best case just query one plugin at all.

Reply:
Moved redundant code into new method.